tests/harness/src/runner.rs

```diff
@@ -0,0 +1,136 @@
+use std::{
+    path::{Path, PathBuf},
+    process::Command,
+    time::Instant,
+};
+
+use anyhow::{Context, Result};
+use ceresdb_client_rs::client::Client;
+use tokio::{
+    fs::{remove_file, File, OpenOptions},
+    io::AsyncWriteExt,
+};
+use walkdir::WalkDir;
+
+use crate::case::TestCase;
+
+/// Input SQL query cases
+const TEST_CASE_EXTENSION: &'static str = "sql";
+/// Output generated by running queries
+const OUTPUT_FILE_EXTENSION: &'static str = "out";
+/// Expected result provides with queries
+const RESULT_FILE_EXTENSION: &'static str = "result";
+/// Tools to compare file
+const DIFF_BINARY: &'static str = "diff";
+
+pub struct Runner {
+    case_root: String,
+    client: Client,
+}
+
+impl Runner {
+    pub fn new<P: AsRef<Path>>(root_dir: P, client: Client) -> Self {
+        Self {
+            case_root: root_dir.as_ref().as_os_str().to_str().unwrap().to_owned(),
+            client,
+        }
+    }
+
+    pub async fn run(&self) {
+        let case_paths = self.collect_cases();
+        let case_count = case_paths.len();
+        let mut diff_cases = vec![];
+        for path in case_paths {
+            let result: Result<()> = try {
+                let case = TestCase::from_file(path.with_extension(TEST_CASE_EXTENSION)).await?;
+                let output_path = path.with_extension(OUTPUT_FILE_EXTENSION);
+                let mut output_file = Self::open_output_file(&output_path).await?;
+
+                let timer = Instant::now();
+                case.execute(&self.client, &mut output_file).await?;
+                let elapsed = timer.elapsed();
+
+                output_file.flush().await?;
+                let is_different = self.compare(&path);
+                if !is_different {
+                    remove_file(output_path).await?;
+                } else {
+                    diff_cases.push(path.as_os_str().to_str().unwrap().to_owned());
```

Comment:
Only collect path, no diff content?

Reply:
Yep, explained above.

One more thing, the whole folder is tracked by git. So we won't miss any ".out" file if there is